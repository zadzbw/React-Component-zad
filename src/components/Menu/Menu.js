/**
 * Created by zad on 17/1/16.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import isOneOf from '../../utils/isOneOf';
import _remove from 'lodash/remove';

const menuPrefix = 'zad-menu';
const _type = {
  mode: ['vertical', 'inline']
};

export default class Menu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      expandKeys: 'expandKeys' in this.props ? this.props.expandKeys : this.props.defaultExpandKeys,
      selectedKeys: 'selectedKeys' in this.props ? this.props.selectedKeys : this.props.defaultSelectedKeys,
    };
    this._expandSub = this::this._expandSub;
    this._selectItem = this::this._selectItem;
  }

  static propTypes = {
    mode: PropTypes.oneOf(['vertical', 'inline']),
    level: PropTypes.number,
    prefix: PropTypes.string,
    expandKeys: PropTypes.arrayOf(PropTypes.string),
    defaultExpandKeys: PropTypes.arrayOf(PropTypes.string),
    selectedKeys: PropTypes.arrayOf(PropTypes.string),
    defaultSelectedKeys: PropTypes.arrayOf(PropTypes.string),
    onExpandChange: PropTypes.func,
    onSelect: PropTypes.func,
    onlyExpandOneSub: PropTypes.bool,
  };

  static defaultProps = {
    mode: 'inline',
    level: 1,
    prefix: menuPrefix,
    defaultExpandKeys: [],
    defaultSelectedKeys: [],
    onExpandChange: () => undefined,
    onSelect: () => undefined,
    onlyExpandOneSub: false,
  };

  componentWillReceiveProps(nextProps) {
    // 为受控组件时，运行如下代码
    const keys = {};
    if ('selectedKeys' in nextProps) {
      keys.selectedKeys = nextProps.selectedKeys;
    }
    if ('expandKeys' in nextProps) {
      keys.expandKeys = nextProps.expandKeys;
    }
    this.setState(keys);
  }

  renderChildren(props) {
    const {expandKeys, selectedKeys} = this.state;
    return React.Children.map(props.children, (child) => {
      return React.cloneElement(child,
        {
          ...child.props,
          level: props.level,
          prefix: props.prefix,
          expandKeys,
          selectedKeys,
          eventKey: child.key,
          _expandSub: this._expandSub,
          _selectItem: this._selectItem,
        });
    });
  }

  _expandSub(args) {
    const expandKeys = this.state.expandKeys.concat();
    const {eventKey, expand, subParentKey} = args;
    if (!expand) {
      // 当onlyExpandOneSub为true时，清空expandKeys数组
      if (!subParentKey && this.props.onlyExpandOneSub) {
        _remove(expandKeys, (key) => eventKey !== key);
      }
      expandKeys.push(eventKey);
    } else {
      _remove(expandKeys, (key) => eventKey === key);
    }
    if (!('expandKeys' in this.props)) {
      this.setState({expandKeys});
    }
    this.props.onExpandChange(expandKeys);
  }

  _selectItem(args) {
    if (!('selectedKeys' in this.props)) {
      this.setState({
        selectedKeys: args.key,
      });
    }
    this.props.onSelect(args);
  }

  render() {
    const {mode} = this.props;
    const menuClass = classNames(menuPrefix, {
      [`${menuPrefix}-${mode}`]: mode && isOneOf(_type.mode, mode),
    });

    return (
      <ul className={menuClass}>
        {this.renderChildren(this.props)}
      </ul>
    );
  }
}
