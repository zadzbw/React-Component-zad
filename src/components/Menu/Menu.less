@import "../../style/variables";
@import "../../style/mixins/index";

@menu-prefix: zad-menu;

.@{menu-prefix} {
  outline: none;
  margin-bottom: 0;
  padding-left: 0;
  list-style: none;
  z-index: @z-index-dropdown;
  color: @text-color;
  background: @body-background;
  line-height: 46px;
  user-select: none;

  &-item-group-title {
    color: @text-color-secondary;
    font-size: @font-size-base;
    line-height: 1.5;
    padding: 8px 16px;
  }

  &-item-group-items {
    margin: 0;
    padding: 0;
  }

  &-item,
  &-sub-title {
    cursor: pointer;
    transition: all 0.3s ease;
    margin: 0;
    padding: 0 20px;
    position: relative;
    display: block;
    .ellipsis();
    &:hover {
      color: @primary-color;
    }
  }

  &-item, &-sub-title, &-item-group-title {
    .fa {
      min-width: 14px;
      margin-right: 0.5em;
    }
  }

  &-inline {
    border-right: 1px solid @border-color-split;
    .@{menu-prefix}-item {
      border-right: 1px solid @border-color-split;
      margin-left: -1px;
      left: 1px;
      position: relative;
      z-index: 1;

      &-selected {
        border-right: 3px solid @primary-color;
        color: @primary-color;
        background-color: tint(@primary-color, 90%);
        transform: translateZ(0);
      }
    }
  }

  &-inline, &-item-group-list {
    & > .@{menu-prefix}-item,
    & > .@{menu-prefix}-sub > .@{menu-prefix}-sub-title {
      padding: 0 16px 0 28px;
      font-size: @font-size-base;
      line-height: 42px;
      height: 42px;
      .ellipsis();
    }
  }

}
