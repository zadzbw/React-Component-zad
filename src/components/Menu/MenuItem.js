/**
 * Created by zad on 17/1/16.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';

export default class MenuItem extends Component {
  render() {
    const {prefix} = this.props;
    const itemClass = classNames({
      [`${prefix}-item`]: true
    });

    return (
      <li className={itemClass}>
        {this.props.children}
      </li>
    );
  }
}
