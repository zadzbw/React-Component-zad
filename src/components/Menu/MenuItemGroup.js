/**
 * Created by zad on 17/1/16.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';

export default class MenuItemGroup extends Component {
  renderChildren(props) {
    return React.Children.map(props.children, (child) => {
      return React.cloneElement(child,
        {
          ...child.props,
          prefix: props.prefix
        });
    });
  }

  render() {
    const {prefix, title} = this.props;
    const groupClass = classNames({
      [`${prefix}-item-group`]: true
    });

    return (
      <li className={groupClass}>
        <div className={`${prefix}-item-group-title`}>
          {title}
        </div>
        <ul className={`${prefix}-item-group-items`}>{this.renderChildren(this.props)}</ul>
      </li>
    );
  }
}
