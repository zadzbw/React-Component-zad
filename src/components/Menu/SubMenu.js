/**
 * Created by zad on 17/1/16.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import Animate from 'rc-animate';
import collapseAnimation from '../../utils/collapseAnimation';

import MenuItem from './MenuItem';
import MenuItemGroup from './MenuItemGroup';
import _SubMenu from './SubMenu';

export default class SubMenu extends Component {
  constructor(props) {
    super(props);
    this.__expandSub = this::this.__expandSub;
    this._isExpand = this::this._isExpand;
  }

  static propTypes = {
    title: PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.element,
    ]),
  };

  static defaultProps = {
    title: '',
  };

  renderChildren(props) {
    return React.Children.map(props.children, (child) => {
      if (child.type !== MenuItem && child.type !== MenuItemGroup && child.type !== _SubMenu) {
        console.error('the children of Menu.SubMenu must to be Menu.Item, Menu.ItemGroup or itself');
        return null;
      }
      return React.cloneElement(child,
        {
          ...child.props,
          // 传递每一级的level，用于缩进
          level: props.level + 1,
          prefix: props.prefix,
          expandKeys: props.expandKeys,
          selectedKeys: props.selectedKeys,
          eventKey: child.key,
          _expandSub: props._expandSub,
          _selectItem: props._selectItem,
          // 用于传递上一级sub的key
          subParentKey: child.type === _SubMenu ? this.props.eventKey : '',
        });
    });
  }

  _isExpand() {
    const {expandKeys, eventKey} = this.props;
    return expandKeys.includes(eventKey);
  }

  __expandSub() {
    const {eventKey, _expandSub, subParentKey} = this.props;
    const param = {
      eventKey,
      expand: this._isExpand(),
      subParentKey,
    };
    _expandSub(param);
  }

  render() {
    const {prefix, title, level} = this.props;
    const expand = this._isExpand();
    const subClass = classNames({
      [`${prefix}-sub`]: true,
      [`${prefix}-sub-expand`]: expand,
    });
    const paddingLeft = level * 20;

    return (
      <li className={subClass}>
        <div className={`${prefix}-sub-title`} onClick={this.__expandSub} style={{paddingLeft}}>
          {title}
        </div>
        <Animate animation={collapseAnimation} component={'ul'}>
          {expand && <ul>{this.renderChildren(this.props)}</ul>}
        </Animate>
      </li>
    );
  }
}
