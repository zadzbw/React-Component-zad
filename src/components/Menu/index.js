/**
 * Created by zad on 17/1/16.
 */
import Menu from './Menu';
import SubMenu from './SubMenu';
import MenuItem from './MenuItem';
import MenuItemGroup from './MenuItemGroup';

import './Menu.less';

Menu.Item = MenuItem;
Menu.SubMenu = SubMenu;
Menu.ItemGroup = MenuItemGroup;

export default Menu;
