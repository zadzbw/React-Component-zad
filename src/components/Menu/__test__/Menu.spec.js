/**
 * Created by zad on 17/1/19.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Menu from '../../Menu/Menu';
import SubMenu from '../../Menu/SubMenu';
import MenuItemGroup from '../../Menu/MenuItemGroup';
import MenuItem from '../../Menu/MenuItem';

describe('Menu Test', () => {
  it('select item', () => {
    const wrapper = shallow(
      <Menu defaultSelectedKeys={['222']} defaultExpandKeys={['sub-1']}>
        <SubMenu key="sub-1" title={'Navigation One'}>
          <MenuItem key="000">00000</MenuItem>
          <MenuItem key="111">11111</MenuItem>
          <MenuItem key="222">22222</MenuItem>
          <MenuItem key="333">33333</MenuItem>
        </SubMenu>
      </Menu>
    );

    // initial
    expect(wrapper.state().selectedKeys).toEqual(['222']);
    expect(wrapper.childAt(0).shallow().find(MenuItem).at(2).shallow().hasClass('zad-menu-item-selected')).toBeTruthy();
    expect(wrapper.childAt(0).shallow().find(MenuItem).at(2).shallow().text()).toEqual('22222');

    // select item 1
    wrapper.childAt(0).shallow().find(MenuItem).at(1).shallow().simulate('click');
    expect(wrapper.state().selectedKeys).toEqual(['111']);
    expect(wrapper.childAt(0).shallow().find(MenuItem).at(1).shallow().hasClass('zad-menu-item-selected')).toBeTruthy();
    expect(wrapper.childAt(0).shallow().find(MenuItem).at(1).shallow().text()).toEqual('11111');

    // select item 3
    wrapper.childAt(0).shallow().find(MenuItem).at(3).shallow().simulate('click');
    expect(wrapper.state().selectedKeys).toEqual(['333']);
    expect(wrapper.childAt(0).shallow().find(MenuItem).at(3).shallow().hasClass('zad-menu-item-selected')).toBeTruthy();
    expect(wrapper.childAt(0).shallow().find(MenuItem).at(3).shallow().text()).toEqual('33333');

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('expand sub', () => {
    const wrapper = shallow(
      <Menu defaultSelectedKeys={['222']} defaultExpandKeys={['sub-1']}>
        <SubMenu key="sub-1" title={'Navigation One'}>
          <MenuItemGroup title={'test1'}>
            <MenuItem key="111">11111</MenuItem>
            <MenuItem key="222">22222</MenuItem>
          </MenuItemGroup>
          <MenuItemGroup title={'test2'}>
            <MenuItem key="333">33333</MenuItem>
            <MenuItem key="444">44444</MenuItem>
          </MenuItemGroup>
        </SubMenu>

        <SubMenu key="sub-3" title={'Navigation Three'}>
          <MenuItem key="option5">Option 5</MenuItem>
          <MenuItem key="option6">Option 6</MenuItem>
          <SubMenu key="sub-3-1" title="Submenu 1">
            <MenuItem key="option7">Option 7</MenuItem>
            <MenuItem key="option8">Option 8</MenuItem>
            <SubMenu key="sub-3-1-1" title="Submenu 2">
              <MenuItem key="option1">Option 1</MenuItem>
              <MenuItem key="option2">Option 2</MenuItem>
            </SubMenu>
          </SubMenu>
        </SubMenu>
      </Menu>
    );

    expect(wrapper.state().expandKeys).toEqual(['sub-1']);

    // child 1 before click
    expect(wrapper.childAt(1).shallow().find(MenuItem).exists()).toBeFalsy();
    expect(wrapper.childAt(1).shallow().find(SubMenu).exists()).toBeFalsy();
    expect(wrapper.childAt(1).shallow().hasClass('zad-menu-sub-expand')).toBeFalsy();

    // child 1 click simulate
    wrapper.childAt(1).shallow().find('.zad-menu-sub-title').simulate('click');

    // child 1 after click
    expect(wrapper.childAt(1).shallow().find(MenuItem).exists()).toBeTruthy();
    expect(wrapper.childAt(1).shallow().find(SubMenu).exists()).toBeTruthy();
    expect(wrapper.childAt(1).shallow().hasClass('zad-menu-sub-expand')).toBeTruthy();
    expect(wrapper.state().expandKeys).toEqual(['sub-1', 'sub-3']);

    // child 0 before click
    expect(wrapper.childAt(0).shallow().find(MenuItem).exists()).toBeTruthy();
    expect(wrapper.childAt(0).shallow().hasClass('zad-menu-sub-expand')).toBeTruthy();

    // child 0 click simulate
    wrapper.childAt(0).shallow().find('.zad-menu-sub-title').simulate('click');

    // child 0 after click
    expect(wrapper.childAt(0).shallow().find(MenuItem).exists()).toBeFalsy();
    expect(wrapper.childAt(0).shallow().hasClass('zad-menu-sub-expand')).toBeFalsy();
    expect(wrapper.state().expandKeys).toEqual(['sub-3']);

    // sub in sub
    wrapper.childAt(1).find(SubMenu).at(0).shallow().find(SubMenu).at(0).shallow().find('.zad-menu-sub-title').simulate('click');
    expect(wrapper.state().expandKeys).toEqual(['sub-3', 'sub-3-1']);
    wrapper.childAt(1).find(SubMenu).at(0).shallow().find(SubMenu).at(0).shallow().find(SubMenu).at(0).shallow().find('.zad-menu-sub-title').simulate('click');
    expect(wrapper.state().expandKeys).toEqual(['sub-3', 'sub-3-1', 'sub-3-1-1']);

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });
});
