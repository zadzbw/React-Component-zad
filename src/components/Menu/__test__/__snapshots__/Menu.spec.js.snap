exports[`Menu Test expand sub 1`] = `
<ul
  className="zad-menu zad-menu-inline">
  <SubMenu
    _expandSub={[Function]}
    _selectItem={[Function]}
    eventKey="sub-1"
    expandKeys={
      Array [
        "sub-3",
        "sub-3-1",
        "sub-3-1-1",
      ]
    }
    level={1}
    prefix="zad-menu"
    selectedKeys={
      Array [
        "222",
      ]
    }
    title="Navigation One">
    <MenuItemGroup
      title="test1">
      <MenuItem>
        11111
      </MenuItem>
      <MenuItem>
        22222
      </MenuItem>
    </MenuItemGroup>
    <MenuItemGroup
      title="test2">
      <MenuItem>
        33333
      </MenuItem>
      <MenuItem>
        44444
      </MenuItem>
    </MenuItemGroup>
  </SubMenu>
  <SubMenu
    _expandSub={[Function]}
    _selectItem={[Function]}
    eventKey="sub-3"
    expandKeys={
      Array [
        "sub-3",
        "sub-3-1",
        "sub-3-1-1",
      ]
    }
    level={1}
    prefix="zad-menu"
    selectedKeys={
      Array [
        "222",
      ]
    }
    title="Navigation Three">
    <MenuItem>
      Option 5
    </MenuItem>
    <MenuItem>
      Option 6
    </MenuItem>
    <SubMenu
      title="Submenu 1">
      <MenuItem>
        Option 7
      </MenuItem>
      <MenuItem>
        Option 8
      </MenuItem>
      <SubMenu
        title="Submenu 2">
        <MenuItem>
          Option 1
        </MenuItem>
        <MenuItem>
          Option 2
        </MenuItem>
      </SubMenu>
    </SubMenu>
  </SubMenu>
</ul>
`;

exports[`Menu Test select item 1`] = `
<ul
  className="zad-menu zad-menu-inline">
  <SubMenu
    _expandSub={[Function]}
    _selectItem={[Function]}
    eventKey="sub-1"
    expandKeys={
      Array [
        "sub-1",
      ]
    }
    level={1}
    prefix="zad-menu"
    selectedKeys={
      Array [
        "333",
      ]
    }
    title="Navigation One">
    <MenuItem>
      00000
    </MenuItem>
    <MenuItem>
      11111
    </MenuItem>
    <MenuItem>
      22222
    </MenuItem>
    <MenuItem>
      33333
    </MenuItem>
  </SubMenu>
</ul>
`;
